src/main/java/org/wildfly/naming/client/util/NamingUtils.java

```diff
@@ -57,7 +57,7 @@ public static CompositeName toCompositeName(final Name name) throws InvalidNameE
             return (CompositeName) name;
         } else {
             final CompositeName compositeName = new CompositeName();
-            compositeName.add(name.toString());
+            compositeName.addAll(new CompositeName(name.toString()));
```

Comment:
The theory behind CompositeName is that each segment is meant to represent a name within a namespace, so it's supposed to be able to support having a single segment that is a name-within-a-schema (e.g. "foo/bar/baz") which is used to navigate to a single naming store.

That said, I will extrapolate that this change means that the naming server is not coping with parsing a name out of a CompositeName component, which is definitely a bug, but one that we can't really grapple with because legacy servers will always have this bug.

So, I think the solution should be to introduce a new method, e.g. toDecomposedCompositeName or something like that, which creates a CompositeName where each name segment is equal to the name segment in the source name.  Connecting to (at minimum) the legacy server should always use this method of conversion, while we still maintain the method which converts a Name to a single-element CompositeName as before.

WDYT?